libraries/plugins/xfpga/usrclk/fpga_user_clk.c

```diff
@@ -0,0 +1,816 @@
+	}
+
+	result = get_usrclk_uio(sysfs_path,
+		0x14,
+		&uio,
+		&uio_ptr);
+	if (result != FPGA_OK) {
+		OPAE_ERR("Failed to get user clock uio");
+		return result;
+	}
+
+	result = usrclk_read_freq(uio_ptr, IOPLL_MEASURE_HIGH, &high);
+	if (result != FPGA_OK) {
+		OPAE_ERR("Failed to get user clock High");
+		opae_uio_close(&uio);
+		return result;
+	}
+
+	result = usrclk_read_freq(uio_ptr, IOPLL_MEASURE_LOW, &low);
+	if (result != FPGA_OK) {
+		OPAE_ERR("Failed to get user clock Low");
+		opae_uio_close(&uio);
+		return result;
+	}
+
+	*userclk_high = high * 10000;	// Adjust to Hz
+	*userclk_low = low * 10000;
+
+	opae_uio_close(&uio);
+	return FPGA_OK;
+}
+
+
+// set fpga user clock
+fpga_result set_userclock(const char *sysfs_path,
+	uint64_t userclk_high,
+	uint64_t userclk_low)
+{
+	char sysfs_usrpath[SYSFS_PATH_MAX] = { 0 };
+	int fd, res, ret                   = 0;
+	char *bufp                         = NULL;
+	size_t cnt                         = 0;
+	uint64_t revision                  = 0;
+	uint8_t seq                        = 1;
+	uint8_t *uio_ptr                   = NULL;
+	fpga_result result                 = FPGA_OK;
+	struct opae_uio uio;
+
+	memset(&uio, 0, sizeof(uio));
+
+	if (sysfs_path == NULL) {
+		OPAE_ERR("Invalid Input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+
+	if (userclk_high < MIN_FPGA_FREQ) {
+		OPAE_ERR("Invalid Input frequency");
+		return FPGA_INVALID_PARAM;
+	}
+
+	// Agilex user clock DFH revision 1
+	// S10 & A10 user clock DFH revision 0
+	result = get_userclk_revision(sysfs_path, &revision);
+	if (result == FPGA_OK && revision == AGILEX_USRCLK_REV) {
+		// Enforce 1x clock within valid range
+		if ((userclk_low > IOPLL_AGILEX_MAX_FREQ) ||
+			(userclk_low < IOPLL_AGILEX_MIN_FREQ)) {
+			OPAE_ERR("Invalid Input frequency");
+			return FPGA_INVALID_PARAM;
+		}
+
+		bufp = (char *)&iopll_agilex_freq_config[userclk_low];
+	} else {
+		// S10 & A10 user clock
+		// Enforce 1x clock within valid range
+		if ((userclk_low > IOPLL_MAX_FREQ) ||
+			(userclk_low < IOPLL_MIN_FREQ)) {
+			OPAE_ERR("Invalid Input frequency");
+			return FPGA_INVALID_PARAM;
+		}
+		bufp = (char *)&iopll_freq_config[userclk_low];
+	}
+
+	ret = using_iopll(sysfs_usrpath, sysfs_path);
+	if (ret == FPGA_OK) {
+
+		fd = open(sysfs_usrpath, O_WRONLY);
+		if (fd < 0) {
+			OPAE_MSG("open(%s) failed: %s",
+				sysfs_usrpath, strerror(errno));
+			return FPGA_NOT_FOUND;
+		}
+
+		cnt = sizeof(struct iopll_config);
+		do {
+			res = write(fd, bufp, cnt);
+			if (res < 0) {
+				OPAE_ERR("Failed to write");
+				break;
+			}
+			bufp += res;
+			cnt -= res;
+		} while (cnt > 0);
+		close(fd);
+
+		return FPGA_OK;
+	} else if (ret == FPGA_NO_ACCESS) {
+		return FPGA_NO_ACCESS;
+	}
+
+	struct pll_config *iopll_config = (struct pll_config *)bufp;
+	if ((iopll_config->pll_freq_khz > IOPLL_MAX_FREQ * 1000) ||
+		(iopll_config->pll_freq_khz < IOPLL_MIN_FREQ * 1000))
+		return FPGA_EXCEPTION;
+
+	result = get_usrclk_uio(sysfs_path,
+		0x14,
```

Comment:
Can we use a #define for the feature id?

Reply:
fixed  added  #define USRCLK_FEATURE_ID 0x14